### DISCUSSION TITLE
Neural-network based SVBRDF and SDFs

### DISCUSSION THREAD

**Discussion by saeedhd96** [mitsuba-renderer/mitsuba3#225] [Q&A] [answered]
Hi,

Thanks for releasing the new version of Mitsuba! 
I was wondering if in this version it is possible to use SVBRDFs in the form of neural networks, i.e. representing BRDF parameters such as albedo and roughness by a neural network that gets location as input.
Additionally I was wondering if it is possible to do ray tracing (sphere tracing) on SDFs using Mitsuba3.

**Comment by Speierers** [comment]
Hi @saeedhd96 ,

Regarding using neural networks within the rendering algorithm (e.g. as a BSDF) is something we are very interested in, although currently it isn't supported. This [PR](https://github.com/mitsuba-renderer/drjit/pull/62) will add support for mixing pytorch and drjit AD pipelines. In theory you could use such custom AD operation to implement a neural BSDF in Mitsuba, but this will force the integrator to operate in wavefront as synchronization will be necessary at every BSDF evaluation.

Regarding sphere tracing, I suggest you take a look at this [repository](https://github.com/rgl-epfl/differentiable-sdf-rendering).

**Reply by saeedhd96** [reply]
Hello,

I see that the PR is merged. Is it now possible to transfer gradients from the mitsuba renderer to a network?

Thanks,
-Saeed

**Reply by maxfrei750** [reply]
Hi! Yes, it is. I'd start with this tutorial: https://mitsuba.readthedocs.io/en/stable/src/inverse_rendering/pytorch_mitsuba_interoperability.html

Have fun!

**Reply by wjakob** [reply]
Just FYI: you cannot call into PyTorch from within a megakernel. Wavefront-style evaluation is your only option in this case. Sébastien's patch will go a long way into making this easier to use, but it does mean that you lose all of the benefits of megakernel-style compilation (which is sort of the point of Dr.Jit/Mitsuba 3).

The main goal of his work is to make it easier to embed Mitsuba 3 into a PyTorch computation. In other words: it allows you to conveniently add neural nets before the renderer or after the renderer. It's putting neural nets *inside* the renderer that is difficult, because we cannot magically "merge" PyTorch/Tensorflow kernels into a megakernel being compiled by Dr.Jit.

**Comment by saeedhd96** [comment]
Thanks for the update and the tutorial.
My question is, right now, the brdf parameters are defined for each object.
In a neural network based brdf, we need to define the brdf parameters as a
function of location on the object. Is it possible to pass such vectors as
brdf params and evaluate it?

Thanks,
-Saeed

**Reply by Speierers** [reply]
This is already what the `Bitmap` texture plugin does in some ways. For instance you can define your own "procedural" texture plugin that will return different value based on the location of the shading point on the geometry (e.g. UV coordinates).

As Wenzel mentioned, if you want to use a neural network to represent this function, you will be forced to either discretize/pre-compuate that function and store it in a table or switch to wavefront mode path tracing which will be considerable slower and use more VRAM.

**Comment by asprasan** [comment]
The paper [Neural BRDF](http://www0.cs.ucl.ac.uk/staff/A.Sztrajman/webpage/publications/nbrdf2021/nbrdf.html) now integrates neural BRDF evaluation into mitsuba. While they provide a C++ plugin, it's not meant for mitsuba3 unfortunately.

But, do I understand correctly, that if I had a pre-trained MLP then I could quite easily write a python plugin for evaluating BRDF from this MLP? I suppose it should take advantage of the Dr. Jit acceleration as I do not require gradients to be passed on to the neural network.

**Reply by saeedhd96** [reply]
Even if you have a pre-trained network, you still cannot use the megakernel when you use a neural network, as far as I understand.
Btw, we have implemented a python imllementation for neural representation of BRDF parameters in Mitsuba3, and we will be releasing our code in less than a week, if that is useful.

**Reply by asprasan** [reply]
Well, I too don't understand fully.
But, I will be looking forward to your code release. It's great news for me personally.
And thank you very much.

**Reply by njroussel** [reply]
Let me shortly expand on [this comment](https://github.com/mitsuba-renderer/mitsuba3/discussions/225#discussioncomment-3665094):

Dr.Jit is constructing graphs of its own computations. Ideally, we'd want to transform these graphs into (few) large kernels (i.e "megakernels"). In fact, with Mitsuba, an entire rendering is one single kernel. 
Introducing something external to Dr.Jit like a PyTorch computation will split the graph because Dr.Jit just simply doesn't know what's happening anymore. Other than some loss of performance, there's functionally nothing wrong with this.  However, because Mitsuba/Dr.Jit were **intentionally** built to produce megakernels it requires some configuration to allow this splitting to happen gracefully. The following two lines are usually enough:
```
dr.set_flag(dr.JitFlag.VCallRecord, False)
dr.set_flag(dr.JitFlag.LoopRecord, False)
```
This will turn-off megakernel mode and produce a wavefront-style computation.

**Reply by asprasan** [reply]
Thanks for the explanation. It does make a lot more sense.
But I thought that for something like MLPs, we might be able to run the forward pass in Dr. Jit (as matrix multiplications/dot products) instead of bringing Pytorch into the mix.
But I am not 100% confident if that's possible. I have to look into it.
But, thanks again for the explanation

**Reply by njroussel** [reply]
Yes, this is/would be possible (and is most likely what @saeedhd96 implemented ?).

**Reply by saeedhd96** [reply]
I wasn't able to do this. This would require to implement neural networks (with or without backward prop) in drjit, which I couldn't do because I didn't find the available matrix multiplication operations in drjit enough. In fact I think it was impossible at the current state of drjit because it only supports matrix multiplication operation for 4x4 at most (Matrix4f, correct? Also, .matmul() doesn't work on TensorXf()).

What we have implemented is in pytorch, and requires disabling the megakernel, and supports both forward/backward pass.

**Reply by njroussel** [reply]
It takes some work (it's not implemented), but definitely possible - someone shared a matrix multiplication routine with `TensorXf` objects [here](https://github.com/mitsuba-renderer/mitsuba3/discussions/579#discussioncomment-5652452).

Anyway, looking forward to your PyTorch implementation :smile:

**Reply by asprasan** [reply]
Hi @saeedhd96 , just a friendly follow-up on your implementation. Has it been released? If not, where should I keep an eye for its release? Thanks in advance.

**Reply by saeedhd96** [reply, marked as answer]
Hi,

I made a notebook for this specific topic, please see #718.
